app_mesh.tf

```
@@ -0,0 +1,203 @@
+                        data.aws_acmpca_certificate_authority.root_ca.arn
+                      ]
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    listener {
+      health_check {
+        protocol            = local.app_mesh_proto
+        path                = lookup(var.health_check, "path", null)
+        healthy_threshold   = 2
+        unhealthy_threshold = 2
+        timeout_millis      = 2000
+        interval_millis     = 5000
+      }
+      port_mapping {
+        port     = var.container_port
+        protocol = local.app_mesh_proto
+      }
+      tls {
+        certificate {
+          acm {
+            certificate_arn = data.aws_acm_certificate.internal.arn
+          }
+        }
+        mode = "STRICT"
+      }
+    }
+    logging {
+      access_log {
+        file {
+          path = "/dev/stdout"
+        }
+      }
+    }
+    service_discovery {
+      aws_cloud_map {
+        namespace_name = "apps.local"
+        service_name   = var.service_name
+      }
+    }
+  }
+}
+
+resource "aws_appmesh_route" "grpc" {
+  for_each = lookup(var.app_mesh_configuration, "grpc_endpoints", {})
+
+  mesh_name           = local.mesh_name
+  name                = "${var.service_name}-${each.key}"
+  virtual_router_name = var.app_mesh_configuration.virtual_router_name
+
+  spec {
+    # 1000 is default (lowest) priority, valid: 0 ~ 1000
+    priority = lookup(each.value, "priority", 1000)
+    grpc_route {
+
+      action {
+        weighted_target {
+          virtual_node = aws_appmesh_virtual_node.this[0].name
+          weight       = 100
+        }
+      }
+      dynamic match {
+        for_each = lookup(each.value, "match", [{
+          /*match must be present, even when empty*/
+        }])
+        content {
+          method_name  = lookup(match.value, "method_name", null)
+          service_name = lookup(match.value, "service_name", null)
+          dynamic metadata {
+            for_each = lookup(match.value, "metadata", {})
+            content {
+              name   = lookup(metadata.value, "name", "missing_name")
+              invert = lookup(metadata.value, "invert", false)
+              match {
+                exact = lookup(metadata.value.match, "exact", null)
+                regex = lookup(metadata.value.match, "regex", null)
+              }
+            }
+          }
+        }
+      }
+
+      timeout {
+        idle {
+          unit  = "ms"
+          value = 5000
+        }
+        per_request {
+          unit  = "s"
+          value = 30
+        }
+      }
+      retry_policy {
+        max_retries       = 3
+        grpc_retry_events = ["cancelled", "unavailable"]
+        per_retry_timeout {
+          unit  = "ms"
+          value = 5000
+        }
+      }
+    }
+  }
+}
+
+resource "aws_appmesh_route" "http" {
+  for_each = lookup(var.app_mesh_configuration, "http_endpoints", {})
```

```suggestion
  for_each = var.app_mesh_configuration.enabled ? lookup(var.app_mesh_configuration, "http_endpoints", {}) : {}
```

to support usages without an app mesh

---

did you test that? I've often found errors stating that `true` / `false` must yield the same type. one is object with members, the other is not.